ludwig/datasets/__init__.py

```diff
@@ -121,12 +123,81 @@ def load_dataset_uris(
     return dataset_out, training_set_out, validation_set_out, test_set_out
 
 
-def _load_cacheable_dataset(dataset: str, backend: Backend) -> CacheableDataframe:
-    dataset_name = dataset[len(URI_PREFIX) :]
-    loader = get_dataset(dataset_name)
-    df = loader.load(split=False)
-    df = backend.df_engine.from_pandas(df)
-    return CacheableDataframe(df=df, name=dataset, checksum=str(loader.get_mtime()))
+@DeveloperAPI
+def _load_hf_datasets(
+    dataset: Optional[Union[str, DataFrame]],
+    training_set: Optional[Union[str, DataFrame]],
+    validation_set: Optional[Union[str, DataFrame]],
+    test_set: Optional[Union[str, DataFrame]],
+    backend: Backend,
+) -> Tuple[
+    Optional[CacheableDataframe],
+    Optional[CacheableDataframe],
+    Optional[CacheableDataframe],
+    Optional[CacheableDataframe],
+]:
+    """Loads and returns any Hugging Face datasets as CacheableDataframes.
+
+    Returns the input unmodified for any non-HF datasets.
+    """
+    dataset_out = dataset
+    training_set_out = training_set
+    validation_set_out = validation_set
+    test_set_out = test_set
+
+    # Check that any of the datasets begin with the `hf://` prefix denoting a Hugging Face dataset URI
+    # Hugging Face datasets should follow the naming convention `hf://<dataset_name>--<dataset_subsample>`
+    if dataset is not None:
+        if isinstance(dataset, str) and dataset.startswith(HF_PREFIX):
+            dataset_out = _load_cacheable_dataset(dataset, backend, hf=True)
+    elif training_set is not None:
+        train_df = test_df = val_df = None
+        training_set_checksum = None
+        if isinstance(training_set, str) and training_set.startswith(HF_PREFIX):
+            # For the training set, we only want to use the TRAINING split of the dataset
+            loader = get_dataset("hugging_face")
+            dataset_name, dataset_subsample = _get_hf_dataset_and_subsample(training_set)
+            setattr(loader.config, "huggingface_dataset_id", dataset_name)
+            setattr(loader.config, "huggingface_subsample", dataset_subsample)
+            train_df, val_df, test_df = loader.load(split=True)
```

Comment:
In the blocks below, we have to use different loaders for the validation and test sets in case the user wants to use validation and test sets from different Hugging Face datasets. We unpack val_df/test_df directly if they come from the same dataset as train_df. Otherwise, we have to load them in using `_load_cacheable_dataset`.

I'm not quite sure what you mean by, "How does the loader know what the training set is?"